pbj-core/pbj-grpc-helidon/src/main/java/com/hedera/pbj/grpc/helidon/PbjProtocolHandler.java

```diff
@@ -0,0 +1,590 @@
+            // connection, tear down resources, etc. It will also forward this on to the client, causing the connection
+            // to be torn down.
+            incoming.onError(e);
+        }
+    }
+
+    /**
+     * Close the connection with the client. May be called by different threads concurrently.
+     */
+    private synchronized void close() {
+        // Canceling a future that has already completed has no effect. So by canceling here, we are saying:
+        // "If you have not yet executed, never execute. If you have already executed, then just ignore me".
+        // The "isCancelled" flag is set if the future was canceled before it was executed.
+        deadlineFuture.cancel(false);
+        currentStreamState = Http2StreamState.HALF_CLOSED_LOCAL;
+    }
+
+    /**
+     * Helper function. Given a string of digits followed by a unit, schedule a callback to be invoked when the
+     * deadline is exceeded, and return the associated future. The proper format of the string is defined in the
+     * specification as:
+     *
+     * <pre>
+     *      Timeout → "grpc-timeout" TimeoutValue TimeoutUnit
+     *      TimeoutValue → {positive integer as ASCII string of at most 8 digits}
+     *      TimeoutUnit → Hour / Minute / Second / Millisecond / Microsecond / Nanosecond
+     *             Hour → "H"
+     *             Minute → "M"
+     *             Second → "S"
+     *             Millisecond → "m"
+     *             Microsecond → "u"
+     *             Nanosecond → "n"
+     * </pre>
+     *
+     * <p>Illegal values result in the deadline being ignored.</p>
+     *
+     * @param timeout The timeout value. Cannot be null.
+     * @return The future representing the task that will be executed if/when the deadline is reached.
+     */
+    private ScheduledFuture<?> scheduleDeadline(@NonNull final String timeout) {
+        final var matcher = GRPC_TIMEOUT_PATTERN.matcher(timeout);
+        if (matcher.matches()) {
+            final var num = Integer.parseInt(matcher.group(1));
+            final var unit = matcher.group(2);
+            final var deadline = System.nanoTime() * TimeUnit.NANOSECONDS.convert(num, switch (unit) {
+                case "H" -> TimeUnit.HOURS;
+                case "M" -> TimeUnit.MINUTES;
+                case "S" -> TimeUnit.SECONDS;
+                case "m" -> TimeUnit.MILLISECONDS;
+                case "u" -> TimeUnit.MICROSECONDS;
+                case "n" -> TimeUnit.NANOSECONDS;
+                // This should NEVER be reachable, because the matcher would not have matched.
+                default -> throw new GrpcException(GrpcStatus.INTERNAL, "Invalid unit: " + unit);
+            });
+            return deadlineDetector.scheduleDeadline(deadline, () -> {
+                route.deadlineExceededCounter().increment();
+                incoming.onError(new GrpcException(GrpcStatus.DEADLINE_EXCEEDED));
+            });
+        }
+
+        return new NoopScheduledFuture();
```

Comment:
This means the given `timeout` is invalid, indicating that the client is "bad". On one hand, it would be nice to log something or emit a metric. On the other hand, doing so could worsen the effects of DoS attacks. So, I'm unsure. Thoughts?